operator/v1/types_ingress.go

```diff
@@ -225,6 +232,97 @@ var (
 	DNSReadyIngressConditionType = "DNSReady"
 )
 
+// SecuritySpec defines the settings for securing IngressController connections.
+type SecuritySpec struct {
+	// Profile is one of three possible security configurations from [1]:
+	//
+	// [1] https://wiki.mozilla.org/Security/Server_Side_TLS#Recommended_configurations
+	//
+	// If unset, the default is "Intermediate".
+	//
+	// +optional
+	Profile *SecurityProfileType `json:"profile,omitempty"`
+	// CustomSettings defines the schema for customizing the security settings of
+	// IngressController connections.
+	//
+	// +optional
+	CustomSettings *CustomProfileSettings `json:"customSettings,omitempty"`
+}
+
+// SecurityProfileType defines a security profile.
+type SecurityProfileType string
+
+const (
+	// SecurityProfileOldType maps to:
+	// https://wiki.mozilla.org/Security/Server_Side_TLS#Old_backward_compatibility
+	SecurityProfileOldType SecurityProfileType = "Old"
+	// SecurityProfileIntermediateType maps to:
+	// https://wiki.mozilla.org/Security/Server_Side_TLS#Intermediate_compatibility_.28default.29
+	SecurityProfileIntermediateType SecurityProfileType = "Intermediate"
+	// SecurityProfileModernType maps to:
+	// https://wiki.mozilla.org/Security/Server_Side_TLS#Modern_compatibility
+	SecurityProfileModernType SecurityProfileType = "Modern"
+)
+
+// CustomProfileSettings defines the schema for a custom security profile.
+type CustomProfileSettings struct {
+	// Ciphers is used to specify the cipher algorithms that are negotiated
+	// during the SSL/TLS handshake with an IngressController. Each cipher must
+	// be an explicit, colon-delimited list of ciphers.
+	//
+	// If unset, the "Intermediate" ciphersuite [1] is used:
+	//
+	// [1] https://wiki.mozilla.org/Security/Server_Side_TLS#Intermediate_compatibility_.28default.29
+	//
+	// +optional
+	Ciphers *string `json:"ciphers,omitempty"`
+	// SecurityProtocol is used to specify one or more encryption protocols
+	// that are negotiated during the SSL/TLS handshake with the IngressController.
+	//
+	// If unset, TLSv1.0, TLSv1.1, TLSv1.2 and TLSv1.3 are used.
+	//
+	// +optional
+	SecurityProtocol *SecurityProtocol `json:"securityProtocol,omitempty"`
+}
+
+// SecurityProtocol defines one or more security protocols used by
+// an IngressController to secure network connections.
+type SecurityProtocol struct {
```

Comment:
How about hedging against future change here and making this one a union type, like:

```go
type SecurityProtocolPolicy struct {
  Type SecurityProtocolPolicyType

  Range *RangeSecurityProtocolPolicy
  Whitelist *WhitelistSecurityProtocolPolicy // possible future policy based on a set
}

type RangeSecurityProtocolPolicy {
  Min *SecurityProtocolVersion
  Max *SecurityProtocolVersion
}
```

Reply:
@ironcladlou thanks for reviewing. I would like to think about your recommendation in more detail before making the change. What are the different `SecurityProtocolPolicyType`s do you envision? I am cautious about making the api more complex than needed. I believe the common use case for `SecurityProtocol` is to simply specify a minimum protocol version to use.

Reply:
I listed a potential future protocol policy in my example, `Whitelist`. Re: complexity, from the user's side, the difference would look like:

```yaml
security:
  protocolPolicy:
    range:
      min:
      max:
```
(note that the discriminator [can be optional](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190325-unions.md), I think)

 vs.

```yaml
security:
  protocolPolicy:
    min:
    max:
```

Reply:
@ironcladlou can you help me better understand your proposal for the `Type` field?

Reply:
This PR proposes:

```yaml
security:
  protocolPolicy:
    min:
    max:
```

What I am proposing:
```yaml
security:
  protocolPolicy:
    type: range
    min:
    max:
```

The union type's purpose being to allow for future alternatives to `range`, such as a theoretical `whitelist` policy:

```yaml
security:
  protocolPolicy:
    type: whitelist
    versions:
    - TLSv1.1
    - TLSv1.2
```

Maybe `range` can already cover all possible practical use cases of `whitelist`. I don't have another example in mind yet — I just wanted to call out the alternative for discussion.

Reply:
I appreciate the discussion topic. IMO I see the use for a simple protocol policy that specifies a single min, a single max, or a min/max pair. I'm drawing a blank as I try to envision other policy types. Let's see if other reviewers chime-in.